shell/platform/common/cpp/client_wrapper/include/flutter/encodable_value.h

```diff
@@ -6,16 +6,150 @@
+ public:
+  // Rely on std::variant for most of the constructors/operators.
+  using super = std::variant<std::monostate,
+                             bool,
+                             int32_t,
+                             int64_t,
+                             double,
+                             std::string,
+                             std::vector<uint8_t>,
+                             std::vector<int32_t>,
+                             std::vector<int64_t>,
+                             std::vector<double>,
+                             EncodableList,
+                             EncodableMap>;
+  using super::super;
+  using super::operator=;
+
+  explicit EncodableValue() : super() {}
+
+  // Avoid the C++17 pitfall of conversion from char* to bool. Should not be
+  // needed for C++20.
+  explicit EncodableValue(const char* string) : super(std::string(string)) {}
+  EncodableValue& operator=(const char* other) {
+    *this = std::string(other);
+    return *this;
+  }
+
+  // Override the conversion constructors from std::variant to make them
+  // explicit, to avoid implicit conversion.
+  //
+  // While implicit conversion can be convenient in some cases, it can have very
+  // surprising effects. E.g., calling a function that takes an EncodableValue
+  // but accidentally passing an EncodableValue* would, instead of failing to
+  // compile, go through a pointer->bool->EncodableValue(bool) chain and
+  // silently call the function with a temp-constructed EncodableValue(true).
+  template <class T>
+  constexpr explicit EncodableValue(T&& t) noexcept : super(t) {}
+
+  // Returns true if the value is null. Convenience wrapper since unlike the
+  // other types, std::monostate uses aren't self-documenting.
+  bool IsNull() const { return index() == 0; }
+
+  // Convience method to simplify handling objects received from Flutter where
+  // the values may be larger than 32-bit, since they have the same type on the
+  // Dart side, but will be either 32-bit or 64-bit here depending on the value.
+  //
+  // It is a programming error to call this method if the value doesn't contain
+  // either an int32_t or an int64_t.
+  int64_t LongValue() {
+    if (std::holds_alternative<int32_t>(*this)) {
+      return std::get<int32_t>(*this);
+    }
+    return std::get<int64_t>(*this);
+  }
+
+  // Convenience wrappers for nested collection types, to allow readable
+  // chaining when not doing type validation, keeping the [] lookup next to the
+  // type extraction. E.g.:
+  //   EncodableValue inner_value =
+  //       value.ListValue()[0].MapValue()[EncodableValue("some_key")];
+  // rather than
+  //   EncodableValue inner_value = std::get<EncodableMap>(
+  //      std::get<EncodableList>(value)[0])[EncodableValue("some_key")];
+  //
+  // It is a programming error to call these method if the value doesn't
+  // contain the correct collection type.
+  const EncodableList& ListValue() const {
```

Comment:
Are these really more convenient? Seems like unnecessary ceremony around std::variant whose uses are well known and documented already. Your call.

Reply:
It's specifically for the chained use called out in the comment. The issue is that the more you chain, the more the type extraction and the indexing operator are separated.

I did originally intend not to have any of these methods, but encountered a couple of places where I was doing things like in the comment, and the new version felt significantly less readable. There's an argument that maybe using temporary variables is a better solution, but it becomes more verbose quickly.

Although those cases were mostly in unit tests, so maybe it's not something that will realistically happen very often. I'll look again at the FDE plugins and see if I ever actually used this in real code, and if not I can remove them under the principle that we can add them later if there is real-world demand, but removing them later is hard.